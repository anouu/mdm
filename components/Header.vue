<template>
  <header class="header">
    <NuxtLink to="/" class="link title-font">M</NuxtLink>
    <div class="actions">
      <template v-if="settingsHidden">
        <button
          class="button button--small"
          title="Syncroniser le panier"
          @click="handleSync"
        >
          <div class="material-icons icon">file_download</div>
        </button>
        <button
          class="button button--small"
          title="Changer de mode"
          @click="handleDarkMode"
        >
          <div class="material-icons icon">
            {{ isDarkMode ? 'light_mode' : 'dark_mode' }}
          </div>
        </button>
      </template>

      <button
        class="button settings"
        title="Ouvrir les paramÃ¨tres"
        @click="settingsHidden = !settingsHidden"
      >
        <div class="material-icons icon">settings</div>
      </button>

      <button
        v-if="!hideBasketButton"
        class="button basket"
        title="Afficher le mini panier"
        @click="handleBasket"
      >
        <div class="material-icons icon">shopping_cart</div>
        <div v-if="!isCartEmpty" class="badge">{{ getCartQty }}</div>
      </button>
    </div>
  </header>
</template>

<script>
import { mapState, mapGetters, mapActions } from 'vuex'

export default {
  props: {
    hideBasketButton: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      settingsHidden: false,
    }
  },

  computed: {
    ...mapState('ui', ['isDarkMode']),
    ...mapGetters(['getCartQty', 'isCartEmpty']),
  },

  methods: {
    ...mapActions(['setSampleCart']),
    ...mapActions('ui', ['setDarkMode', 'setBasketMini']),

    handleDarkMode() {
      this.setDarkMode(!this.isDarkMode)
    },

    handleBasket() {
      this.setBasketMini(true)
    },

    async handleSync() {
      const delay = Math.floor(Math.random() * (800 - 300) + 300)
      await this.$tools.sleep(delay)

      await this.setSampleCart()
    },
  },
}
</script>

<style lang="scss" scoped>
.header,
.settings,
.actions {
  display: flex;
  align-items: center;
}

.header {
  position: fixed;
  inset: 0 0 auto 0;
  background-color: var(--light-color);
  justify-content: space-between;
  padding: var(--spacing-small) var(--spacing);
  color: var(--dark-color);
  border-bottom: 4px solid var(--brand-color-2);
  z-index: 20;
}

.link {
  font-size: clamp(24px, calc(5vw + 1rem), 90px);
  color: var(--brand-color-1);
}

.icon {
  font-size: var(--icon-size);
  height: var(--icon-size);
  width: var(--icon-size);
}

.button {
  line-height: 0;

  --icon-size: clamp(24px, calc(2vw + 1rem), 40px);

  font-size: var(--icon-size);

  &--small {
    --icon-size: clamp(16px, calc(1vw + 1rem), 30px);
  }

  & + & {
    margin-left: 0.5em;
  }
}

.badge {
  background-color: var(--light-color);
  position: absolute;
  inset: rem(-14) rem(-14) auto auto;
  width: rem(22);
  height: rem(22);
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  color: var(--brand-color-1);
  font-size: rem(14);
  font-weight: 700;
  line-height: 0;
  border-radius: 50%;
  border: 2px solid var(--brand-color-2);
}
</style>
