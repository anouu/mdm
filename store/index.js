export const state = () => ({
  cart: {},
})

export const getters = {
  getCartItems: (state) => {
    return state.cart?.products || []
  },

  getCartQty: (state, getters) => {
    return Object.values(getters.getCartItems).reduce(
      (acc, item) => acc + item.qty,
      0
    )
  },

  getCartTotal: (state, getters) => {
    return Object.values(getters.getCartItems).reduce(
      (acc, item) => acc + item.qty * item.price.base.amount,
      0
    )
  },

  isCartEmpty: (state, getters) => {
    return getters.getCartQty === 0
  },
}

export const mutations = {
  SET_CART(state, newCart) {
    state.cart = newCart
  },

  UPDATE_PRODUCT(state, { reference, updatedItem }) {
    state.cart.products[reference] = updatedItem
  },
}

export const actions = {
  async setSampleCart({ commit }) {
    try {
      const { cart } = await fetch('/api/sampleCart').then((res) => res.json())

      commit('SET_CART', cart)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
    }
  },

  updateProduct({ commit, getters }, { reference, qty }) {
    const items = getters.getCartItems
    const updatedItem = { ...items[reference], qty }

    commit('UPDATE_PRODUCT', { reference, updatedItem })
  },

  deleteProduct(context, reference) {
    const items = context.getters.getCartItems
    const newCart = { ...context.state.cart, products: items }

    delete items[reference]

    context.commit('SET_CART', newCart)
    context.commit('ui/SET_MODAL_OPEN', false)
  },
}
