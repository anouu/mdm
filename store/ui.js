export const state = () => ({
  isBackdropOpen: false,
  isModalOpen: false,
  isBasketMiniOpen: false,
  isDarkMode: false,
  modalData: {
    status: false,
    data: {
      body: '',
      action: null,
      actionData: null,
    },
  },
})

export const mutations = {
  SET_BACKDROP_OPEN(state, boolean) {
    state.isBackdropOpen = boolean
  },

  SET_MODAL_OPEN(state, boolean) {
    state.isModalOpen = boolean
    state.isBackdropOpen = boolean
  },

  SET_DRAWER_OPEN(state, boolean) {
    state.isBasketMiniOpen = boolean
    state.isBackdropOpen = boolean
  },

  CLOSE_ALL_OPEN(state) {
    state.isBackdropOpen = false
    state.isModalOpen = false
    state.isBasketMiniOpen = false
  },

  SET_MODAL_DATA(state, data) {
    state.modalData = data
  },

  SET_DARK_MODE(state, boolean) {
    state.isDarkMode = boolean
  },
}

export const actions = {
  closeAll({ commit }) {
    commit('CLOSE_ALL_OPEN', false)
  },

  setBasketMini({ commit }, boolean) {
    commit('SET_DRAWER_OPEN', boolean)
  },

  setModal({ commit }, { status, data }) {
    commit('SET_MODAL_DATA', data)
    commit('SET_MODAL_OPEN', status)
  },

  setDarkMode({ commit }, boolean) {
    commit('SET_DARK_MODE', boolean)
  },
}
