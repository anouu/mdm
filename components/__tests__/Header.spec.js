import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Vuex from 'vuex'
import * as Ui from '../../store/ui'
import Header from '../Header'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('Header', () => {
  let actions
  let getters
  let store
  let factory
  let actionsUi

  beforeEach(() => {
    getters = {
      getCartQty: () => 0,
      isCartEmpty: () => true,
    }
    actions = {
      setSampleCart: jest.fn(),
    }
    actionsUi = {
      setBasketMini: jest.fn(),
    }
    store = new Vuex.Store({
      getters,
      actions,
      modules: {
        ui: {
          state: Ui.state,
          actions: actionsUi,
          namespaced: true,
        },
      },
    })

    factory = () => {
      return shallowMount(Header, {
        store,
        localVue,
        stubs: {
          NuxtLink: RouterLinkStub,
        },
      })
    }
  })

  it('calls HandleBasket', () => {
    const wrapper = factory()
    const handleBasket = jest.spyOn(wrapper.vm, 'handleBasket')

    wrapper.vm.handleBasket()

    expect(handleBasket).toHaveBeenCalled()
  })

  test('mounts properly', () => {
    const wrapper = factory()
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
