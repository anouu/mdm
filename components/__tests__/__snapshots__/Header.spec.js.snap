// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders properly 1`] = `
<header class="header"><a class="link title-font">M</a>
  <div class="actions">
    <!----> <button title="Ouvrir les paramÃ¨tres" class="button settings">
      <div class="material-icons icon">settings</div>
    </button> <button title="Afficher le mini panier" class="button basket">
      <div class="material-icons icon">shopping_cart</div>
      <!---->
    </button>
  </div>
</header>
`;
